subiquity/controllers/snaplist.py

```diff
@@ -0,0 +1,236 @@
+            cb(self.model.get_snap_list())
+
+    def get_snap_list(self, callback):
+        if self.state >= SnapInfoLoaderState.LOADING_INFO:
+            callback(self.model.get_snap_list())
+        elif None in self.ongoing:
+            self.ongoing[None].append(callback)
+        else:
+            self.start()
+            self.ongoing[None].append(callback)
+
+    def get_snap_info(self, snap, callback):
+        if len(snap.channels) > 0:
+            callback()
+            return
+        if snap in self.ongoing:
+            self.ongoing[snap].append(callback)
+            return
+        if snap in self.pending_info_snaps:
+            self.pending_info_snaps.remove(snap)
+        self._fetch_info_for_snap(snap, callback)
+
+    def _fetch_info_for_snap(self, snap, callback):
+        self.ongoing[snap] = [callback]
+        log.debug('starting fetch for %s', snap.name)
+        self.run_in_bg(
+            lambda: self._bg_fetch_next_info(snap),
+            lambda fut: self._fetched_info(snap, fut))
+
+    def _fetch_next_info(self):
+        if not self.pending_info_snaps:
+            self.state = SnapInfoLoaderState.LOADED
+            return
+        snap = self.pending_info_snaps.pop(0)
+        self._fetch_info_for_snap(snap, self._fetch_next_info)
+
+    def _bg_fetch_next_info(self, snap):
+        return self.session.get(self.url_base + 'name=' + snap.name, timeout=60)
+
+    def _fetched_info(self, snap, fut):
+        if self.state == SnapInfoLoaderState.STOPPED:
+            return
+        try:
+            response = fut.result()
+            response.raise_for_status()
+        except requests.exceptions.RequestException:
+            log.exception("loading snap info failed")
+            self.state = SnapInfoLoaderState.FAILED
+            # XXX something better here?
+        else:
+            data = response.json()
+            self.model.load_info_data(data)
+        for cb in self.ongoing.pop(snap):
+            cb()
+
+
+class SnapListController(BaseController):
+
+    signals = [
+        ('network-config-written', 'network_config_done'),
+        ('network-proxy-set', 'proxy_config_done'),
+    ]
+
+    def __init__(self, common):
+        super().__init__(common)
+        self.model = self.base_model.snaplist
+        self.loader = None
+        self._maybe_start_new_loader()
+        self.answers = self.all_answers.get('SnapList', {})
+
+    def _maybe_start_new_loader(self):
+        if self.loader:
+            if self.loader.state >= SnapInfoLoaderState.LOADING_INFO:
+                return
+            else:
+                self.loader.stop()
+        if self.opts.snaps_from_examples:
+            self.loader = SampleDataSnapInfoLoader(
+                self.model,
+                os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "examples", "snaps"))
+        else:
+            self.loader = SnapdSnapInfoLoader(self.model, self.run_in_bg, '/run/snapd.socket')
```

Comment:
the snap socket could be a class attribute or module value.

Reply:
Sure, but this is about where we store hard coded stuff.  Generally we want to see those at the module or class level, rather than buried in a function.